--- alpsize-05-boost/wolff.C	2013-05-01 14:14:04.000000000 +0900
+++ alpsize-06-parameters/wolff.C	2013-05-01 14:14:03.000000000 +0900
@@ -28,6 +28,7 @@
 *
 *****************************************************************************/
 
+#include <alps/parameter.h>
 #include <boost/array.hpp>
 #include <boost/random.hpp>
 #include <boost/timer.hpp>
@@ -36,15 +37,16 @@
 #include <stack>
 #include <vector>
 
-#define L 32
-#define N (L*L)
-#define T 2.2
-#define MCSTEP (1 << 15)
-#define MCTHRM (MCSTEP >> 3)
-#define SEED 93812
-
 int main() {
 
+  alps::Parameters params(std::cin);
+  const int L = params.value_or_default("L", 32);
+  const int N = L * L;
+  const double T = params.value_or_default("T", 2.2);
+  const int MCSTEP = params.value_or_default("SWEEPS", 1 << 15);
+  const int MCTHRM = params.value_or_default("THERMALIZATION", MCSTEP >> 3);
+  const unsigned int SEED = params.value_or_default("SEED", 93812);
+
   // setting up square lattice
   std::vector<boost::array<int, 4> > nn(N);
   for (int y = 0; y < L; ++y)
